File: sprint5/By lessons/lesson8code5/lesson8code5.py
# Функция возвращает новый корень поддерева.
def rotate(vertex):
	if abs(vertex.left.height - vertex.right.height) < 2:
		# Вращать не надо, поддерево с вершиной vertex и так сбалансировано.	
		return vertex
	if (vertex.left.height - vertex.right.height == -2):
		# Нам нужны левые вращения.
		b = vertex.right
		R = b.right
		C = b.left
		
		if C.height <= R.height:
			# Нужно малое левое вращение.
			return small_left_rotation(vertex)
		else:
			# Нужно большое левое вращение.
			return big_left_rotation(vertex)

	if (vertex.left.height - vertex.right.height == 2):
		# Нам нужны правые вращения.
		b = vertex.left
		C = b.right
		L = b.left

		if C.height <= L.height:
			# Нужно малое правое вращение.
			return small_right_rotation(vertex)
		else:
			# Нужно большое правое вращение.
			return big_right_rotation(vertex)

File: sprint5/By languages/swift/lesson8code5.swift
// Функция возвращает новый корень поддерева.
func rotate<T: Comparable>(_ vertex: Node<T>) -> Node<T> {
    if abs(vertex.left.height - vertex.right.height) < 2 {
        // Вращать не надо, поддерево с вершиной vertex и так сбалансировано.
        return vertex
    }
    
    if vertex.left.height - vertex.right.height == -2 {
        // Нам нужны левые вращения.
        let b = vertex.right
        let R = b.right
        let C = b.left
        
        if C.height <= R.height {
            // Нужно малое левое вращение.
            return smallLeftRotation(vertex)
        } else {
            // Нужно большое левое вращение.
            return bigLeftRotation(vertex)
        }
    }
    
    if vertex.left.height - vertex.right.height == 2 {
        // Нам нужны правые вращения.
        let b = vertex.left
        let C = b.right
        let L = b.left
        
        if C.height <= L.height {
            // Нужно малое правое вращение.
            return smallRightRotation(vertex)
        } else {
            // Нужно большое правое вращение.
            return bigRightRotation(vertex)
        }
    }
}
